x-pack/plugins/rule_registry/server/rule_data_client/index.ts

```diff
@@ -70,9 +70,10 @@ export class RuleDataClient implements IRuleDataClient {
     };
   }
 
-  getWriter(options: { namespace?: string } = {}): RuleDataWriter {
+  async getWriter(options: { namespace?: string } = {}): Promise<RuleDataWriter> {
     const { namespace } = options;
     const alias = getNamespacedAlias({ alias: this.options.alias, namespace });
+    await this.rolloverAliasIfNeeded({ namespace });
```

Comment:
Why `rolloverAliasIfNeeded` is called at the point when we get a writer, but index creation (`createWriteTargetIfNeeded`) is deferred to the point when we actually index something, and get an exception saying that the index does not exist?

Can/should we combine the two into a single `rolloverAliasOrCreateInitialIndex`?

Reply:
Yeah we can just pass in the alias directly, good point.

The call to `createWriteTargetIfNeeded` is deferred at @dgieselaar 's request to avoid populating the cluster with more empty indices than necessary. We can defer it since attempting to write to an index that doesn't exist returns an error that we can catch, then create the index and write again. We can't defer the call to `rolloverAliasIfNeeded` because there most likely won't be an error to catch if we try to write to an index whose mapping hasn't been updated - for most of the mapping updates we make (e.g. adding a new field to the mapping), it will still be able to write the document but the fields newly added to the template simply won't be mapped in the new documents. By rolling over before returning the writer, we can ensure that when we do write documents the write index has the updated mappings

Reply:
> We can't defer the call to `rolloverAliasIfNeeded` because there most likely won't be an error to catch if we try to write to an index whose mapping hasn't been updated - for most of the mapping updates we make (e.g. adding a new field to the mapping), it will still be able to write the document but the fields newly added to the template simply won't be mapped in the new documents.

👍  yeah, understood that part, agree

As for deferring `createWriteTargetIfNeeded` until the very last moment, I'm personally not sure if this is super necessary because I suspect the time between the moment of getting a writer and attempting to write something will be very short. Exception would be if we get a writer but don't write anything - in this case we'd create the initial index, but it would stay empty for a while.

Both options work for me, I think this is not super important, just thought that maybe combining them into a `rolloverAliasOrCreateInitialIndex` would make the code slightly simpler.